### ISSUE TITLE
Reading a state that was created after the snapshot was taken or in a snapshot that has not yet been applied

### ISSUE THREAD

**Issue by Jolly5** [medizininformatik-initiative/Medication-Graph-FHIR-Converter#13]
In some cases, when attempting to start the graph database population, this error occurs:

```
ERROR GraphDbPopulatorScreenModel:19 - Graph DB population failed.
java.lang.IllegalStateException: Reading a state that was created after the snapshot was taken or in a snapshot that has not yet been applied
	at androidx.compose.runtime.snapshots.SnapshotKt.readError(Snapshot.kt:1930) ~[runtime-desktop-1.6.2.jar:?]
	at androidx.compose.runtime.snapshots.SnapshotKt.current(Snapshot.kt:2275) ~[runtime-desktop-1.6.2.jar:?]
	at androidx.compose.runtime.SnapshotMutableStateImpl.setValue(SnapshotState.kt:308) ~[runtime-desktop-1.6.2.jar:?]
	at de.medizininformatikinitiative.medgraph.ui.theme.templates.ProgressIndicationViewState.setPrimaryTaskDescription(ProgressIndication.kt:105) ~[main/:?]
	at de.medizininformatikinitiative.medgraph.ui.theme.templates.ProgressIndicationViewState.onTaskStackChanged(ProgressIndication.kt:93) ~[main/:?]
	at de.medizininformatikinitiative.medgraph.common.mvc.NamedProgressableImpl.lambda$setTaskStack$0(NamedProgressableImpl.java:28) ~[medgraph-1.0.jar:?]
	at java.util.concurrent.ConcurrentHashMap$KeySetView.forEach(ConcurrentHashMap.java:4709) ~[?:?]
	at de.medizininformatikinitiative.medgraph.common.mvc.Dispatcher.dispatchEvent(Dispatcher.java:51) ~[medgraph-1.0.jar:?]
	at de.medizininformatikinitiative.medgraph.common.mvc.NamedProgressableImpl.setTaskStack(NamedProgressableImpl.java:27) ~[medgraph-1.0.jar:?]
	at de.medizininformatikinitiative.medgraph.graphdbpopulator.GraphDbPopulation.executeDatabasePopulation(GraphDbPopulation.java:59) ~[medgraph-1.0.jar:?]
	at de.medizininformatikinitiative.medgraph.ui.desktop.graphdbpopulator.GraphDbPopulatorScreenModel.runPopulationTaskChain(GraphDbPopulatorScreenModel.kt:115) ~[main/:?]
	at de.medizininformatikinitiative.medgraph.ui.desktop.graphdbpopulator.GraphDbPopulatorScreenModel.populateSync(GraphDbPopulatorScreenModel.kt:93) ~[main/:?]
	at de.medizininformatikinitiative.medgraph.ui.desktop.graphdbpopulator.GraphDbPopulatorScreenModel.access$populateSync(GraphDbPopulatorScreenModel.kt:25) ~[main/:?]
	at de.medizininformatikinitiative.medgraph.ui.desktop.graphdbpopulator.GraphDbPopulatorScreenModel$populate$1.invokeSuspend(GraphDbPopulatorScreenModel.kt:76) ~[main/:?]
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33) ~[kotlin-stdlib-1.9.23.jar:1.9.23-release-779]
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:104) ~[kotlinx-coroutines-core-jvm-1.8.0.jar:?]
	at kotlinx.coroutines.internal.LimitedDispatcher$Worker.run(LimitedDispatcher.kt:111) ~[kotlinx-coroutines-core-jvm-1.8.0.jar:?]
	at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:99) ~[kotlinx-coroutines-core-jvm-1.8.0.jar:?]
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:585) ~[kotlinx-coroutines-core-jvm-1.8.0.jar:?]
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:802) ~[kotlinx-coroutines-core-jvm-1.8.0.jar:?]
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:706) ~[kotlinx-coroutines-core-jvm-1.8.0.jar:?]
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:693) ~[kotlinx-coroutines-core-jvm-1.8.0.jar:?]
```

Please fix this :)

**Comment by Jolly5** [comment]
Okay, I think what's happening is as follows:
The `GraphDbPopulatorScreenModel` asynchronously creates the `GraphDbPopulation` and launches the population.

The now available `GraphDbPopulation` triggers a recomposition, which calls into the `ProgressIndication` function, which generates a `ProgressIndicationViewState`. This view state creates its attributes and registers itself as listener on the `GraphDbPopulationObject`. Now, **this happens in a composable function, meaning it's running within a snapshot. The attributes of the view state are generated within this snapshot!**.
Now, if the composition exits and then the `GraphDbPopulation` starts sending updates to that created view state, everything is fine. But if that first update happens before the composition exits, the attributes of the `ProgressIndicationViewState` only exist within the composition's snapshot. However, the `GraphDbPopulation`, which runs asynchronously and is therefore only linked to the global snapshot, attempts to update these attributes. But the attributes do not yet exist in the global snapshot! This then causes the shown exception.

Solution: Do not create the `ProgressIndicationViewState` from the composition. Let the view model handle it.

**Comment by Jolly5** [comment]
Should be fixed as of 5020c5651ad5f53fd6bcc19ad0437eb063b29152.
